### PR TITLE
Updated CategoryFilter to prevent NPE with Cucumber and JUnit Categories

### PR DESCRIPTION
**Pull request by andyrbell**
Fix for #1510 

- NullPointerException was being thrown if JUnit Categories were used in a project containing Cucumber tests.
- Ensure CategoryFilter does not call createSuiteDescription if the description test class is null
- Added integration test to reproduce the error

This bug was originally reported by @krosenvold on the [Gradle Forum](https://discuss.gradle.org/t/npe-with-excludecategories-and-cucumber/13969).

Any of the checked boxes below indicate that I took action:

- [x] Reviewed the [Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md).
- [x] Signed the [Gradle CLA](http://gradle.org/contributor-license-agreement/).
- [x] Ensured that basic checks pass: `./gradlew quickCheck`

For all non-trivial changes that modify the behavior or public API:

- [ ] Before submitting a pull request, I started a discussion on the [Gradle developer list](https://groups.google.com/forum/#!forum/gradle-dev),
      the [forum](https://discuss.gradle.org/) or can reference a [JIRA issue](https://issues.gradle.org/secure/Dashboard.jspa).
- [ ] I considered writing a design document. A design document can be
brief but explains the use case or problem you are trying to solve,
touches on the planned implementation approach as well as the test cases
that verify the behavior. Optimally, design documents should be submitted
as a separate pull request. [Samples](https://github.com/gradle/gradle/tree/master/design-docs)
can be found in the Gradle GitHub repository. Please let us know if you need help with
creating the design document. We are happy to help!
- [x] The pull request contains an appropriate level of unit and integration
test coverage to verify the behavior. Before submitting the pull request
I ran a build on my local machine via the command
`./gradlew quickCheck <impacted-subproject>:check`.
- [ ] The pull request updates the Gradle documentation like user guide,
DSL reference and Javadocs where applicable.

### CODE REVIEW

File: subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec/supportsCategoriesAndNullTestClassDescription/build.gradle [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    mavenCentral()
+}
+
+dependencies {
+    testCompile "junit:junit:4.12"
+}
+
+test {
+    useJUnit {
+        excludeCategories 'CategoryA'
```
Comment (bmuschko) [inline subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec/supportsCategoriesAndNullTestClassDescription/build.gradle]: What's the relevance of this category if it is excluded anyway?

File: subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec/supportsCategoriesAndNullTestClassDescription/src/test/java/DescriptionWithNullClassTest.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+@RunWith(CustomRunner.class)
+public class DescriptionWithNullClassTest {
+
+    @Test
+    public void someTest() throws Exception {
```
Comment (bmuschko) [inline subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec/supportsCategoriesAndNullTestClassDescription/src/test/java/DescriptionWithNullClassTest.java]: Doesn't need to throw an exception.

File: subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec/supportsCategoriesAndNullTestClassDescription/src/test/java/DescriptionWithNullClassTest.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
```
Comment (bmuschko) [inline subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec/supportsCategoriesAndNullTestClassDescription/src/test/java/DescriptionWithNullClassTest.java]: Please remove the unnecessary new lines.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec.groovy [hunk 4]
```diff
@@ -64,4 +64,16 @@ public class JUnitCategoriesIntegrationSpec extends AbstractIntegrationSpec {
         def result = new DefaultTestExecutionResult(testDirectory)
         result.testClass("org.gradle.SomeTest").assertTestFailed("initializationError", startsWith("org.gradle.api.GradleException: JUnit Categories defined but declared JUnit version does not support Categories."))
     }
+
+    def supportsCategoriesAndNullTestClassDescription() {
+        when:
```
Comment (bmuschko) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec.groovy]: Can you please a new line between `when` and `then`?

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec.groovy [hunk 5]
```diff
@@ -64,4 +64,16 @@ public class JUnitCategoriesIntegrationSpec extends AbstractIntegrationSpec {
...
+        run "test"
+        then:
+        ":test" in nonSkippedTasks
+        and:
```
Comment (bmuschko) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec.groovy]: Let's remove the last `and` statement. It's OK to just have this under `then`.

File: subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec/supportsCategoriesAndNullTestClassDescription/build.gradle (hunk 1 above)
Comment (andyrbell) [inline subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec/supportsCategoriesAndNullTestClassDescription/build.gradle]: `CategoryFilter`, the source of the bug, is only invoked if a category inclusion / exclusion is specified here. It doesn't matter what is excluded. In fact, no test is even annotated with this category.

File: subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec/supportsCategoriesAndNullTestClassDescription/src/test/java/DescriptionWithNullClassTest.java (hunk 2 above)
Comment (andyrbell) [inline subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec/supportsCategoriesAndNullTestClassDescription/src/test/java/DescriptionWithNullClassTest.java]: Done

File: subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec/supportsCategoriesAndNullTestClassDescription/src/test/java/DescriptionWithNullClassTest.java (hunk 3 above)
Comment (andyrbell) [inline subprojects/testing-jvm/src/integTest/resources/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec/supportsCategoriesAndNullTestClassDescription/src/test/java/DescriptionWithNullClassTest.java]: Done

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec.groovy (hunk 4 above)
Comment (andyrbell) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec.groovy]: Done

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec.groovy (hunk 5 above)
Comment (andyrbell) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/junit/JUnitCategoriesIntegrationSpec.groovy]: Done

**Comment by bmuschko** [comment]
Thanks for the pull request. It has been merged!

**Comment by andyrbell** [comment]
Awesome, thanks!

**Comment by bmuschko** [comment]
Sure thing. Sorry about the long wait.

**Pull request merged**